pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,556 @@
+	installedCIImport    cache.Indexer
+	installedGWName      string
+	namespace            string
+}
+
+func NewMCRouteController(
+	mcClient mcclientset.Interface,
+	gwInformer mcinformers.GatewayInformer,
+	ciImportInformer mcinformers.ClusterInfoImportInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCRouteController {
+	controller := &MCRouteController{
+		mcClient:             mcClient,
+		ovsBridgeClient:      ovsBridgeClient,
+		ofClient:             client,
+		interfaceStore:       interfaceStore,
+		nodeConfig:           nodeConfig,
+		gwInformer:           gwInformer,
+		gwLister:             gwInformer.Lister(),
+		gwListerSynced:       gwInformer.Informer().HasSynced,
+		ciImportInformer:     ciImportInformer,
+		ciImportLister:       ciImportInformer.Lister(),
+		ciImportListerSynced: ciImportInformer.Informer().HasSynced,
+		queue:                workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "gatewayroute"),
+		installedCIImport:    cache.NewIndexer(clusterInfoImportKeyFunc, cache.Indexers{}),
+		namespace:            namespace,
+	}
+	controller.gwInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	controller.ciImportInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueClusterInfoImport(cur)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueueClusterInfoImport(cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueClusterInfoImport(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func clusterInfoImportKeyFunc(obj interface{}) (string, error) {
+	return obj.(*ciImportInfo).ciImport.Name, nil
+}
+
+func (c *MCRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("received unexpected object: %v", obj)
+			return
+		}
+		gw, ok = deletedState.Obj.(*mcv1alpha1.Gateway)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-Gateway object: %v", deletedState.Obj)
+			return
+		}
+	}
+	if !isDelete {
+		if !(gw.InternalIP == "" || gw.GatewayIP != "") {
+			klog.Errorf("no valid Internal IP or GatewayIP is found in Gateway %s", gw.Namespace+"/"+gw.Name)
+			return
+		}
+	}
+	if net.ParseIP(gw.InternalIP) == nil || net.ParseIP(gw.GatewayIP) == nil {
```

Comment:
This should not be checked for deletion too? Then we might just replace line161 with this single check?

Reply:
yes, I refined it and it's single check now.